File: EMR_BMED.Backend/Controllers/UserController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

using EMR_BMED.Backend.Exceptions;
using EMR_BMED.Backend.Models;
using EMR_BMED.Backend.Services;
using EMR_BMED.Backend.Utils;

namespace EMR_BMED.Backend.Controllers
{
  [ApiController]
  [Route("[controller]")]
  public class UserController(UserService userService) : ControllerBase
  {
    [HttpGet("{id}")]
    public async Task<IActionResult> GetOne([FromRoute] Guid id)
    {
      try
      {
        return Ok(await userService.GetOneAsync(id));
      }
      catch (UserNotFoundException ex)
      {
        return NotFound(ex.Message);
      }
    }

    [HttpGet("search/{query}")]
    public IActionResult SearchPatients([FromRoute] string query)
    {
      return Ok(userService.SearchPatients(query));
    }

    [Authorize]
    [HttpPut("patient/{id}")]
    public async Task<IActionResult> Update(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id, [FromBody] PatientUpdateDto dto)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != id)
      {
        return Forbid();
      }

      try
      {
        await userService.UpdateAsync(id, dto);
        return Ok();
      }
      catch (UserNotFoundException ex)
      {
        return NotFound(ex.Message);
      }
      catch (EmailIsTakenException ex)
      {
        return Conflict(new { email = ex.Message });
      }
    }

    [Authorize]
    [HttpPut("doctor/{id}")]
    public async Task<IActionResult> Update(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id, [FromBody] DoctorUpdateDto dto)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != id)
      {
        return Forbid();
      }

      try
      {
        await userService.UpdateAsync(id, dto);
        return Ok();
      }
      catch (UserNotFoundException ex)
      {
        return NotFound(ex.Message);
      }
      catch (EmailIsTakenException ex)
      {
        return Conflict(new { email = ex.Message });
      }
    }

    [Authorize]
    [HttpDelete("{id}")]
    public async Task<IActionResult> Delete(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != id)
      {
        return Forbid();
      }

      try
      {
        await userService.DeleteAsync(id);
        return NoContent();
      }
      catch (UserNotFoundException ex)
      {
        return NotFound(ex.Message);
      }
    }
  }
}

File: EMR_BMED.Backend.JS/controllers/user.controller.js
import express from "express";
const router = express.Router();

import userService from "../services/user.service.js";
import tokenUtils from "../utils/token.utils.js";

router
  .get("patient/all", (req, res) => {
    const patients = userService.getAllPatients();
    return res.status(200).send(patients);
  })
  .get("/search/:query", (req, res) => {
    const patients = userService.searchPatients(req.params.query);
    return res.status(200).send(patients);
  })
  .get("/:id", (req, res) => {
    try {
      const user = userService.getOne(req.params.id);
      return res.status(200).send(user);
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .put("/patient/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== id) {
      return res.status(403).send({});
    }

    try {
      userService.updateOne(id, req.body);
      return res.status(200).send({});
    } catch (err) {
      if (err.name === "UserNotFoundError") {
        return res.status(404).send({ message: err.message });
      } else if (err.name === "EmailIsTakenError") {
        return res.status(409).send({ email: err.message });
      }
    }
  })
  .put("/doctor/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== id) {
      return res.status(403).send({});
    }

    try {
      userService.updateOne(id, req.body);
      return res.status(200).send({});
    } catch (err) {
      if (err.name === "UserNotFoundError") {
        return res.status(404).send({ message: err.message });
      } else if (err.name === "EmailIsTakenError") {
        return res.status(409).send({ email: err.message });
      }
    }
  })
  .delete("/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== id) {
      return res.status(403).send({});
    }
    try {
      userService.deleteOne(id);
      return res.status(204).send({});
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  });

export default router;
